KAFKA-7895: fix Suppress changelog restore

vvcephei opened the pull request:
Several issues have come to light since the 2.2.0 release:

* upon restore, `suppress` incorrectly set the record metadata using the changelog record, instead of preserving the original metadata
* restoring a tombstone incorrectly didn't update the buffer size and min-timestamp

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-7895: fix Suppress changelog restore

vvcephei commented:
Hi @ableegoldman @cadonna @guozhangwang ,

Please review these bugfixes for Suppression when you get the chance.

The reporter has confirmed it fixes the duplicates in his repro as well: https://issues.apache.org/jira/browse/KAFKA-7895?focusedCommentId=16810770&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16810770

I'm still chasing down a different bug to do with the interaction between Suppression and EOS, but I don't think it's related and should be a separate PR.

Thanks,
-John

guozhangwang commented:
Thanks @vvcephei for the PR. Do you want me to review it now or wait for some further investigations?

vvcephei commented:
Sorry for the delay in responding, @guozhangwang . I think we should review this now.

Right now, it seems like the EOS thing is a different mechanism, so I plan to do a separate PR once I figure it out.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -79,17 +84,17 @@ public Headers headers() {
-        long size = 0L;
-        size += 8; // value.context.timestamp
-        size += 8; // value.context.offset
+    public int sizeBytes() {
```
Why this change from `long` to `int`?

vvcephei replied:
`long` is an unnecessarily large data type for `size`, since arrays can only be indexed with integers (thus, a byte array cannot hold more than a "max int" number of bytes).

vvcephei replied:
Actually, the latter conversation with @mjsax about the relationship (or lack of one) between this and the serialized form makes me think maybe this response was incorrect. I'll move it back to long.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -79,17 +84,17 @@ public Headers headers() {
         if (headers != null) {
             for (final Header header : headers) {
-                size += header.key().toCharArray().length;
+                size += header.key().getBytes(UTF_8).length;
```
Is there a chance here that the cumulative size of bytes for all headers and values exceeds max int?  I realize the probability is very low, but should we have a check here and throw an exception in case of an overflow (maybe I'm overthinking this?)

vvcephei replied:
That's a good idea. I agree it's unlikely, and if it does happen, we'd wind up with an exception while serializing later, but we could detect it sooner here and throw a nicer error.

mjsax replied:
I doubt that we could exceed it. Note, that `max.message.byte` is an integer broker side config, and I believe it applies to the serialized version of a message. \cc @guozhangwang @hachikuji Is this correct?

Of course, a user could emit a new key-value pairs with huge key/value/header arrays, however, this message could not be stored in Kafka itself either. If we want a guard, we should use `long size` internally here, and check if it exceed `INT_MAXVALUE` and throw of course.

vvcephei replied:
Thanks for the information. I think this is getting a little complicated, and we'll just drop the check. If the value is too big, we won't be able to serialize it later anyway.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -327,8 +390,13 @@ public void evictWhile(final Supplier<Boolean> predicate,
-                    final ContextualRecord value) {
-        cleanPut(time, key, value);
+                    final ContextualRecord contextualRecord) {
+        if (contextualRecord.value() == null) {
```
why not use `Objects.requireNonNull` here and in the next branch?

vvcephei replied:
Only because `requireNonNull` throws a `NullPointerException`, which I felt was less appropriate than an `IllegalArgumentException` in this case.

mjsax replied:
The argument that `NullPointerException` is not appropriate holds for almost all cases when `requireNonNull` in use. Thus, I would prefer code consistency and use `requireNonNull` here, too.

bbejeck commented on streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java:
```diff
@@ -16,15 +16,83 @@
+        private final List<ProducerRecord<byte[], byte[]>> collected = new LinkedList<>();
+
+        @Override
+        public <K, V> void send(final String topic, final K key, final V value, final Headers headers, final Integer partition, final Long timestamp, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
```
nit: args on separate line

vvcephei replied:
oof! Sorry, I missed these long lines.

bbejeck commented on streams/src/test/java/org/apache/kafka/test/MockInternalProcessorContext.java:
```diff
@@ -16,15 +16,83 @@
+        }
+
+        @Override
+        public <K, V> void send(final String topic, final K key, final V value, final Headers headers, final Long timestamp, final Serializer<K> keySerializer, final Serializer<V> valueSerializer, final StreamPartitioner<? super K, ? super V> partitioner) {
```
nit: args on separate line

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -222,60 +230,109 @@ public void flush() {
                 final long time = timeAndValue.getLong();
                 final byte[] value = new byte[record.value().length - 8];
                 timeAndValue.get(value);
+                if (record.headers() == null || record.headers().lastHeader("v") == null) {
```
If we are serializing the `ContextualRecord` could we ever hit this case?

vvcephei replied:
Yep, the currently released versions of Suppress are already serializing changelog messages without this header. The purpose of this check is to be able to transition to new Streams versions in a backward-compatible way.

mjsax replied:
Shouldn't headers be `null` or contain exactly one header with `key=="v"` and non-null value? If, the second part of "or" should never be true?

vvcephei replied:
If it's an "old format" changelog record, the headers would have been copied from the record context, in which case they could be null or non-null, but still not contain a "v". Of course, it could contain a "v" which has nothing to do with our algorithm here, but I think that situation is undetectable. And unlikely, as well.

mjsax replied:
I don't think so. We never write headers in the changelogger. Note, that the changelog topic is used to recover the store content. However, rows in a store only have a key and a value. There is no header that we could write, because the on put, the current record header does not related to the store content.

Similarly, `suppress()` serializes the whole record context and store it in the value IIRC.

vvcephei replied:
good call. So, we're covered. Thanks for this analysis.

mjsax replied:
Well, my point is, that the check can be simplified. I don't think that `record.headers() == null` can be true; it's guaranteed that there is a headers object.

Not sure if we can simplify the second check. It iterators over the headers map and does String comparison to find a header with key `v` -- seems to be rather heavy.

guozhangwang replied:
If I understand correct, a record read from the changelog topic should only be:

1) having no headers at all (old version)
2) having a singleton header with `v --> byte`.

All other cases should indicate a bug in the code.

So it seems we can just check if `record.headers() == null`, and inside the if condition though, we should also check the `v` and assume it's always there (if not then throw illegal state), and switching on the byte value:

1) byte == 1: do as below.
2) otherwise: to not support forward compatibility, we can just throw unsupported.

vvcephei replied:
We were previously setting the changelog record headers from the record headers. I really don't think it's safe to just assume that record headers are always empty for records inside of Streams.

Also, I'd like to point out that @guozhangwang thinks that old-format headers are _always_ null, and @mjsax think that they are _never_ null, but _always_ empty...

The point is, header handling is partially undefined inside Streams. I'd be very much in favor of being defensive in the face of such ambiguity, rather than making potentially faulty assumptions, since the risk is a runtime exception or even data corruption (if we mis-interpret the binary format, but it happens to de-serialize without exceptions into gibberish).

Further, none of this feedback is pointing out a problem with the code, just that there might be one unnecessary clause in the boolean expression (which, I still am not convinced is unnecessary after all).

guozhangwang replied:
Thanks for the explanation @vvcephei , makes sense.

One thing I'd still like to ask though: from the current code it seems we do not check the byte of `v` if it indeed exist, while I think to be safer we should still check it, and throw exception indicating that we do not support forward compatibility when the format has evolved. WDYT?

mjsax replied:
- It's a contract that `KafkaConsumers` _guarantees_ that `header != null`. \cc @hachikuji to confirm.
- And we know that KafkaStreams never writes headers into changelog topics.

Thus, I don't see any reason to check for something that we know is the case, ie, we know that `header.size() == 0` in old format. For everything else, we could throw `IllegalStateException`. Of course the header API is limiting and we cannot get `size()` and thus `record.headers().lastHeader("v") == null)` is what we need to do... :( -- but we can safely remove the first `null` check -- it could even mask a bug and we should rather fail for this case.

We can also do a version check as suggested by Guozhang.

vvcephei replied:
> we know that KafkaStreams never writes headers into changelog topics.

We're restoring records that we wrote like this: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java#L240

We write headers to the suppression changelog topic. Sorry if that wasn't clear, that's why I'm insisting we have to check it.

If there's a contract that headers are never null, then I agree we can remove that check. In that case, we should add null-checks to the ConsumerRecord constructor.

vvcephei replied:
Ok, I've had a chat with @hachikuji , and he confirmed that the headers _should_ never be null. I did a quick audit of code base as well, and agree that this should be the case.

I added a null-check to the header and docs to this effect in f5bd0999c (in this PR), and then removed the null-check.

I also added an exception if the record has a "v" header, but it isn't `1`. I'm wondering, @guozhangwang , if we should just assume in this case that the record is old-format, and happened to already have a "v" header. The alternative thing is you could have run Streams with a _newer_ version, and then downgraded. In that case, it would be wrong to try restoring as old-format, and an exception would be preferable. I'm leaning toward the exception, since people haven't been using Suppress that long, so the risk of a spurious "v" header is low. WDYT?

mjsax replied:
> We're restoring records that we wrote like this: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java#L240
>
> We write headers to the suppression changelog topic. Sorry if that wasn't clear, that's why I'm insisting we have to check it.

Arg. We should not have done this... But serialize the header in KS and add them to the value. Well, too late now I guess. Or could we switch the format? Note sure what the impact on the upgrade path would be? Thought?

> I added a null-check to the header and docs to this effect in f5bd099 (in this PR), and then removed the null-check.

+100

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -222,60 +230,109 @@ public void flush() {
+                             new ContextualRecord(
+                                 value,
+                                 new ProcessorRecordContext(
+                                     record.timestamp(),
```
Unless I'm missing something aren't we restoring the record with the changelog `offset`, `topic` etc?

vvcephei replied:
Yeah, this is the (incorrect) behavior of the current code. Luckily, this will only happen for the very small window of emitting records that are currently buffered when you upgrade Streams to "this" version. Once those are flushed out, the buffer will always be sending/receiving "v1" changelog messages that correctly encode the upstream topic/offset.

I figured the best way to handle this transition is to just preserve the existing behavior for "old format" messages, rather than try to do something clever.

mjsax replied:
I cannot follow. What is this discussion about?

vvcephei replied:
We're setting the record metadata incorrectly here, which is actually the bug I'm fixing with the new format. It's just that there's apparently nothing better to do that continue handling the old changelog format the same way, since the format itself is missing the information we need.

bbejeck reviewed (commented):
@vvcephei I've made a pass and have a few comments. Also I'm thinking we should have an integration test that fails without your changes in this PR.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/ContextualRecord.java:
```diff
@@ -42,6 +43,31 @@ long sizeBytes() {
+    byte[] serialize() {
+        final byte[] serializedContext = recordContext.serialize();
+        final ByteBuffer buffer =
+            ByteBuffer.allocate(serializedContext.length + Integer.BYTES + (value == null ? 0 : value.length));
```
Could you please assign `Integer.BYTES` to a variable with a meaningful name and use that variable here? It would make the code bit clearer.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -222,60 +230,109 @@ public void flush() {
+    private void logValue(final Bytes key, final BufferKey bufferKey, final ContextualRecord value) {
+        final byte[] serializedContextualRecord = value.serialize();
+
+        final byte[] timeAndContextualRecord = ByteBuffer.wrap(new byte[Long.BYTES + serializedContextualRecord.length])
```
Also here, could you please give `Long.BYTES` a meaningful name?

vvcephei commented:
Thanks for the review, @bbejeck . I've addressed your comments and added an integration test that checks that the topic name is right, at least. It seems a little tricky to deterministically verify the incoming event's offset or partition number.

vvcephei commented:
To actually generate duplicates under the condition I fixed, we'd have to run the application for a pretty long time.

guozhangwang commented:
I've only made a brief pass on the bullet points that's listed on the description and these fixes lgtm.

cc @mjsax @ableegoldman to make another thorough pass over it.

vvcephei commented:
Thanks for the review, @cadonna ! I'm currently seeing if I can get the integration test to spit out duplicates. I'll address your comments in the next commit.

vvcephei commented:
Ok @bbejeck and @cadonna ,

I managed to update the integration test to re-create the duplication bug on trunk (you can see this in action if you check out https://github.com/vvcephei/kafka/tree/testing-test , which has this test cherry-picked onto trunk). I've verified that it fails on trunk both because the topic is wrong on suppressed records *and* because it produces duplicates.

Thanks for inducing me to go the extra mile, @bbejeck .

I think this is ready for final reviews.

-John

vvcephei commented:
Flaky test failure: https://issues.apache.org/jira/browse/KAFKA-7965?focusedCommentId=16815432&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16815432

Retest this, please.

asfgit commented:
SUCCESS 
 10841 tests run, 69 skipped, 0 failed.
--none--

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -79,26 +84,128 @@ public Headers headers() {
                 }
             }
         }
+        if (size < 0) {
```
This does not guard against "double overflow". I think, if we apply a check, we should check for all corner cases or not check at all.

vvcephei replied:
Is there a way to check against wrapping all the way around to positive again? Even a long could wrap around in principle... This is starting to get a little complicated. Maybe we should just fall back on the jvm to catch us when we try to make and populate the byte array after all.

mjsax replied:
> Is there a way to check against wrapping all the way around to positive again?

I think you would need to check after each modification... This would of course become costly.

> Even a long could wrap around in principle... 

True. But I think the risk would be minimal in practice for this case.

> Maybe we should just fall back on the jvm to catch us when we try to make and populate the byte array after all.

Not sure what you mean by this?

vvcephei replied:
I meant that maybe we should just remove the check.

mjsax replied:
> or not check at all.

Fine with we to remove the check.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -79,26 +84,128 @@ public Headers headers() {
 
+    public byte[] serialize() {
+        final byte[] topicBytes = topic == null ? null : topic.getBytes(UTF_8);
+        final byte[][] headerKeys;
```
nit: `headerKey` -> `headerKeyBytes` (same for `headerValues` below)

mjsax replied:
This comment is acked but not addressed. Did it slip? Or did you decide to ignore it?

vvcephei replied:
Oy, it slipped. Fixing it now.

vvcephei replied:
I shouldn't have acked it until I fixed it; that's why I overlooked it on later passes.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -79,26 +84,128 @@ public Headers headers() {
+        int size = 0;
+        size += Long.BYTES; // value.context.timestamp
+        size += Long.BYTES; // value.context.offset
         if (topic != null) {
```
How could `topic` be `null`?

vvcephei replied:
I could see making an assumption that a specific ConsumerRecord is always has a topic (because that's where they come from), but this is the ProcessorRecordContext. The class has no null checks on the topic field, so rather than tracing all the usages to find out whether it can ever be null, in main code or test code, it's simpler and bulletproof just to check for it.

mjsax replied:
`ProcessorRecordContext` would only set the topic to `null` for calls into `init()`, `close()`, and punctuation callbacks. Thus, it should never be `null` here IMHO and I believe it would indicate a bug, if it's `null` here. Hence, this check might masked a bug and we should rather fail with a NPE for this case. \cc @guozhangwang to confirm.

vvcephei replied:
Ok, so I have a clarification: this method is used to estimate the size of the record context in memory, for a variety of usages not restricted to processing. Including some of the cases where you point out it may be null, so we do need the null-check in this method.

In the de//serialize methods below, I've removed the null-check, following the spirit of your feedback.

mjsax replied:
I understand your argument. However, I am not 100% sure if this does not introduce a weird asymmetry? Also, we only estimate the size _if_ we serialize the context. We don't estimate in-memory/object space usage, do we?

vvcephei replied:
Yes, we do. This is also used for computing the size of the record cache to determine cached state store flushing, etc. (I was a little surprised when a bunch of "unrelated" tests failed as a result of the change I tried here).

I shared your concern about the asymmetry, but I think it's actually ok, even if you only consider the suppression buffer. The `sizeBytes` method is an estimation of the resident size of this object in memory, which is only loosely related to the size of the serialized form of the data in this object.

mjsax replied:
Ack. Thanks for being patient with me :)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/ContextualRecord.java:
```diff
@@ -42,6 +43,35 @@ long sizeBytes() {
+        final ProcessorRecordContext context = ProcessorRecordContext.deserialize(buffer);
+
+        final int valueLength = buffer.getInt();
+        final byte[] value = new byte[valueLength];
```
`valueLength` could be `-1`

vvcephei replied:
good call!

guozhangwang replied:
This seems not fixed yet?

vvcephei replied:
Gah. Thanks for the catch!

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -222,60 +238,112 @@ public void flush() {
                 final byte[] value = new byte[record.value().length - 8];
                 timeAndValue.get(value);
+                if (record.headers() == null || record.headers().lastHeader("v") == null) {
+                    cleanPut(time, key,
```
nit: move `key` to new line

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -298,14 +366,20 @@ public void evictWhile(final Supplier<Boolean> predicate,
 
             // predicate being true means we read one record, call the callback, and then remove it
             while (next != null && predicate.get()) {
+                if (next.getKey().time != minTimestamp) {
```
Why do we need this check?

vvcephei replied:
This is one of the bugs that this PR fixes. The correct operation of this class depends on maintaining `minTimestamp` on every mutation. If a future code change breaks that contract, it would re-introduce a variant of our bug, and it would be just as subtle as it was this time. This check safeguards against that eventuality, by checking on our invariant at the spot we depend on it.

mjsax replied:
Ack. So we don't need it, but we want it :)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java:
```diff
@@ -99,31 +108,26 @@ public SuppressionDurabilityIntegrationTest(final boolean eosEnabled) {
+                input,
+                Consumed.with(STRING_SERDE, STRING_SERDE))
+            .groupByKey()
+            .count(Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as(storeName).withCachingDisabled());
```
Why this change?

vvcephei replied:
When I wrote the tests the first time around, I overlooked a simpler application to test suppression. Switching to this simpler application made it possible to directly test for the bug (duplicated suppress output).

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java:
```diff
@@ -179,17 +190,17 @@ public void shouldRecoverBufferAfterShutdown() {
-                )
+                new HashSet<>(asList(
+                    new KeyValueTimestamp<>("k4", 1L, scaledTime(4L)),
+                    new KeyValueTimestamp<>("k5", 1L, scaledTime(5L))
```
Why change from `v` to `k` ?

vvcephei replied:
Instead of grouping on the values anymore, now I'm just counting events. I just didn't realize that simpler aggregation was available when I wrote the tests the first time.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java:
```diff
@@ -256,13 +327,13 @@ private long scaledTime(final long unscaledTime) {
     }
 
-    private void produceSynchronously(final String topic, final List<KeyValueTimestamp<String, String>> toProduce) {
+    private static void produceSynchronouslyToPartition0(final String topic, final List<KeyValueTimestamp<String, String>> toProduce) {
```
nit: `produceSynchronouslyToPartition0` -> `produceSynchronouslyToPartitionZero`

mjsax reviewed (commented):
Did not review all test yet

vvcephei commented:
Thanks for the reviews, @bbejeck, @cadonna, and @mjsax . I've addressed all comments. Do you mind taking another look?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -222,60 +230,109 @@ public void flush() {
+                        memBufferSize -= computeRecordSize(bufferKey.key, removed);
+                    }
+                    if (bufferKey.time == minTimestamp) {
+                        final BufferKey leastKey = sortedMap.firstKey();
```
Just reviewing this PR, it reminds me on this JIRA: https://issues.apache.org/jira/browse/KAFKA-8235

Seems `firstKey` would throw if `sortedMap` is empty.

vvcephei replied:
Oof! You're absolutely right. What a strange api asymmetry...

vvcephei committed:
revert changes to ProcessorContext size estimation

vvcephei commented:
Ok, in adopting the PR comments regarding assuming the topic will never be null, I broke a whole bunch of tests. I've fixed it with the following strategy:

* the topic may be null during size estimation
* the topic will never be null when serializing the record context

vvcephei commented:
Java 8 failed with flaky ConsumerBounceTest. This test was purportedly fixed after my build started, so I'm giving it the benefit of the doubt and not reporting it (https://issues.apache.org/jira/browse/KAFKA-7965).

Retest this, please.

vvcephei commented:
Hey, the tests passed! Do you think we're ready to merge this, @mjsax ?

guozhangwang reviewed (commented):
@vvcephei I made a pass, and seems there are still a couple of minor questions open. Will leave them to you to reply or address, and leave to @mjsax or @bbejeck to merge afterwards.

bbejeck commented:
@vvcephei I've made another pass LGTM.  Just ping us when the last bits are addressed and we can merge

vvcephei committed:
fix

vvcephei commented:
Thanks, all. @guozhangwang and @mjsax , I believe the only remaining unresolved thread is regarding the header check above. Thanks!

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorRecordContext.java:
```diff
@@ -99,6 +104,111 @@ public long sizeBytes() {
+        final long offset = buffer.getLong();
+        final int topicSize = buffer.getInt();
+        final String topic;
+        {
```
Why this block?

vvcephei replied:
It just prevents polluting the method scope with the temp variable `topicBytes`

vvcephei committed:
cr feedback

vvcephei committed:
cr comment

vvcephei committed:
cr feedback: header handling

vvcephei commented:
Ok, @guozhangwang and @mjsax , I think I've addressed your feedbacks. I also did a sweep for any more missed feedback (sorry about that).

Please take another look when you get the chance (also @bbejeck and @cadonna ). Thanks for the reviews!

mjsax reviewed (approved):
I just looked at the last commits only. LGTM now.

mjsax commented:
Java11 failed. Flaky tests tracked in Jira. Java8 passed.

Retest this please.

bbejeck commented:
Merged #6536 into trunk.

bbejeck commented:
Thanks for the fix @vvcephei!
